Update virtual machine entities to latest spec

paulmey opened the pull request:
Also cleaned up the code a bit. Preparing another PR for moving the utility methods to a different package.

ahmetalpbalkan commented on management/virtualmachineimage/entities.go:
```diff
@@ -4,6 +4,7 @@ import (
 	"encoding/xml"
 
 	"github.com/MSOpenTech/azure-sdk-for-go/management"
+	disk "github.com/MSOpenTech/azure-sdk-for-go/management/virtualmachinedisk"
```
It might be better to call it vmdisk

ahmetalpbalkan commented on management/virtualmachinedisk/entities.go:
```diff
@@ -6,3 +6,12 @@ import "github.com/MSOpenTech/azure-sdk-for-go/management"
 	client management.Client
 }
+
+type HostCachingType string // see "HostCachingType*"" constants
```
Comments not really necessary as it's obvious :-) and thats not really a godoc comment either

ahmetalpbalkan commented on management/virtualmachine/entities.go:
```diff
@@ -178,14 +287,24 @@ type RoleSize struct {
 
+type IPAddressType string // see "IPAddressType*"" constants
+
+const IPAddressTypePrivate = IPAddressType("Private") // Only allowed value (currently) for IPAddressType
```
Use const(....) block and "const ( xxx IPAddrType = "foo" ) should be ok, i.e. no need cast type

ahmetalpbalkan commented on management/virtualmachine/entities.go:
Move the type next to identifier, type conversion will be inferred automatically

paulmey commented:
Fixed all comments

ahmetalpbalkan commented on management/virtualmachine/client.go:
```diff
@@ -648,8 +648,8 @@ func (self VirtualMachineClient) createServiceCertDeploymentConf(certPath string
 
-func (self VirtualMachineClient) createSshConfig(certPath, userName string) (SSH, error) {
-	sshConfig := SSH{}
+func (self VirtualMachineClient) createSshConfig(certPath, userName string) (*SSH, error) {
```
why not return SSH and reference with `&` at the caller?

paulmey replied:
That would need an extra variable at the call site [1]. This code is being deleted in #65 , so I don't really care. LMK if you still want me to make the change.

[1] https://github.com/paulmey/azure-sdk-for-go/blob/virtualmachine-cleanup/management/virtualmachine/client.go#L607-607

ahmetalpbalkan replied:
no big deal. having all these pointers is just annoying. I can understand if we re using them for omitting things in XML marshaling but otherwise we shouldnt be doing any pointers, ideally. right?

ahmetalpbalkan replied:
just looked at the link. please do introduce the extra var. :)

paulmey replied:
Totally agree and yes, they are only here to make xml.Marshall work properly... :-(

Sent from my Windows Phone
________________________________
From: Ahmet Alp Balkan
Sent: 3/14/2015 2:38 PM
To: MSOpenTech/azure-sdk-for-go
Cc: Paul Meyer
Subject: Re: [azure-sdk-for-go] Update virtual machine entities to latest spec (#64)


In management/virtualmachine/client.go<https://github.com/MSOpenTech/azure-sdk-for-go/pull/64#discussion_r26442000>:

> @@ -648,8 +648,8 @@ func (self VirtualMachineClient) createServiceCertDeploymentConf(certPath string
>       return certConfig, nil
>  }
>
> -func (self VirtualMachineClient) createSshConfig(certPath, userName string) (SSH, error) {
> -     sshConfig := SSH{}
> +func (self VirtualMachineClient) createSshConfig(certPath, userName string) (*SSH, error) {


no big deal. having all these pointers is just annoying. I can understand if we re using them for omitting things in XML marshaling but otherwise we shouldnt be doing any pointers, ideally. right?

-
Reply to this email directly or view it on GitHub<https://github.com/MSOpenTech/azure-sdk-for-go/pull/64/files#r26442000>.

paulmey replied:
Ok

Sent from my Windows Phone
________________________________
From: Ahmet Alp Balkan
Sent: 3/14/2015 2:38 PM
To: MSOpenTech/azure-sdk-for-go
Cc: Paul Meyer
Subject: Re: [azure-sdk-for-go] Update virtual machine entities to latest spec (#64)


In management/virtualmachine/client.go<https://github.com/MSOpenTech/azure-sdk-for-go/pull/64#discussion_r26442007>:

> @@ -648,8 +648,8 @@ func (self VirtualMachineClient) createServiceCertDeploymentConf(certPath string
>       return certConfig, nil
>  }
>
> -func (self VirtualMachineClient) createSshConfig(certPath, userName string) (SSH, error) {
> -     sshConfig := SSH{}
> +func (self VirtualMachineClient) createSshConfig(certPath, userName string) (*SSH, error) {


just looked at the link. please do introduce the extra var. :)

-
Reply to this email directly or view it on GitHub<https://github.com/MSOpenTech/azure-sdk-for-go/pull/64/files#r26442007>.

ahmetalpbalkan commented on management/virtualmachine/client.go:
```diff
@@ -665,7 +665,7 @@ func (self VirtualMachineClient) createSshConfig(certPath, userName string) (SSH
 	publicKey.Fingerprint = fingerprint
 	publicKey.Path = "/home/" + userName + "/.ssh/authorized_keys"
```
do you think we can just accept certPath as input to this method? hardcoded home path wouldn't work on windows or mac.

paulmey replied:
This is the path where the authorized_keys file is on the vm being deployed. Not the certificate. That should already exist in the cloud service...

Sent from my Windows Phone
________________________________
From: Ahmet Alp Balkan
Sent: 3/14/2015 2:27 PM
To: MSOpenTech/azure-sdk-for-go
Cc: Paul Meyer
Subject: Re: [azure-sdk-for-go] Update virtual machine entities to latest spec (#64)


In management/virtualmachine/client.go<https://github.com/MSOpenTech/azure-sdk-for-go/pull/64#discussion_r26441925>:

> @@ -665,7 +665,7 @@ func (self VirtualMachineClient) createSshConfig(certPath, userName string) (SSH
>       publicKey.Fingerprint = fingerprint
>       publicKey.Path = "/home/" + userName + "/.ssh/authorized_keys"


do you think we can just accept certPath as input to this method? hardcoded home path wouldn't work on windows or mac.

-
Reply to this email directly or view it on GitHub<https://github.com/MSOpenTech/azure-sdk-for-go/pull/64/files#r26441925>.

ahmetalpbalkan replied:
ah ok

paulmey commented:
removed pointer to SSH in return from method

ahmetalpbalkan commented on management/virtualmachine/entities.go:
```diff
@@ -11,132 +12,235 @@ type VirtualMachineClient struct {
-type RoleInstanceList struct {
-	RoleInstance []*RoleInstance
+	UseCertAuth bool   `xml:"-"`
+	CertPath    string `xml:"-"`
```
I think these stuff should not be here unless they're going to be serialized. If these are fields that are used in utility methods, looks like there's a separation of concerns problem here.

paulmey replied:
fixed in #65

ahmetalpbalkan replied:
@paulmey umm but am I supposed to merge like this or not? :) that's why small and atomic commits are always good.

paulmey replied:
Yes, please merge like this. This is not new code, it's the inability of git to match this diff properly. See https://github.com/MSOpenTech/azure-sdk-for-go/blob/master/management/virtualmachine/entities.go#L65-L66 .

I'm not convinced that this PR would have made sense in smaller pieces, but I do agree that small commits are easier to review. Would you have preferred to review this PR in smaller commits and then have it squashed and rebased before merging?

ahmetalpbalkan commented on management/virtualmachine/entities.go:
safe to remove this comment

ahmetalpbalkan commented on management/virtualmachine/entities.go:
belongs to which field? not legit godoc

ahmetalpbalkan commented on management/virtualmachine/entities.go:
try running godoc and see if this comment indeed goes to `Path`

paulmey commented on management/virtualmachine/entities.go:
I'll move it to the type

paulmey commented on management/virtualmachine/entities.go:
moved multiline comments before fields.

ahmetalpbalkan commented:
next time let's limit comments to max 150 lines of diff. it's hard to read like this.

paulmey commented:
Sure, next time I'll break it up into smaller PR's. Thanks

ahmetalpbalkan commented:
@paulmey doesn't have to be separate PR. multi-commit PRs are easy to review since if you'rer refactoring N things not related to each other with different purposes, it makes sense to split it into smaller commits with messages describing the purpose.

ahmetalpbalkan commented on management/virtualmachine/client.go:
```diff
@@ -718,7 +719,7 @@ func (self VirtualMachineClient) createNetworkConfig(os string, sshPort int) (Co
 	endpoint := InputEndpoint{}
 	endpoint.Name = name
-	endpoint.Protocol = protocol
+	endpoint.Protocol = InputEndpointProtocol(protocol)
```
change the signature instead?

paulmey replied:
ah yes, good suggestion

paulmey commented:
Ah ok, I was under the impression that single commit PR's are preferred by most projects. :+1: Thanks for that info, I'll apply it in the future... :-)

ahmetalpbalkan commented on management/virtualmachine/client.go:
```diff
@@ -505,19 +504,18 @@ func (self VirtualMachineClient) addDockerPort(configurationSets []Configuration
 	deployment.DeploymentSlot = "Production"
 	deployment.Label = role.RoleName
-	deployment.RoleList.Role = append(deployment.RoleList.Role, role)
+	deployment.RoleList = append(deployment.RoleList, role)
```
doesn't this append look a bit weird to you?

also we should probably use struct initializer here.

paulmey replied:
changed to struct initializer + some others as well

ahmetalpbalkan commented:
@paulmey compilation fails `management/virtualmachine/client.go:537: undefined: oSVirtualHardDisk
`

paulmey commented:
@ahmetalpbalkan ai...sorry, that should not have happened...

ahmetalpbalkan commented on management/virtualmachine/client.go:
```diff
@@ -461,27 +460,24 @@ func (self VirtualMachineClient) ResolveRoleSize(roleSizeName string) error {
-	return shutdownRoleOperation
+	return ShutdownRoleOperation{
+		OperationType: "ShutdownRoleOperation",
+		Xmlns:         azureXmlns,
```
is this XmlNs thing really necessary to correctly make the request?

paulmey replied:
I don't think so. I'll remove all of them...

ahmetalpbalkan replied:
are we testing this manually by any chance? Am I safe to merge it? :-) 


On Monday, Mar 16, 2015 at 3:49 PM, Paul Meyer <notifications@github.com>, wrote:

In management/virtualmachine/client.go:

>  }
>  
>  func (self VirtualMachineClient) createShutdowRoleOperation() ShutdownRoleOperation {
> -	shutdownRoleOperation := ShutdownRoleOperation{}
> -	shutdownRoleOperation.OperationType = "ShutdownRoleOperation"
> -	shutdownRoleOperation.Xmlns = azureXmlns
> -
> -	return shutdownRoleOperation
> +	return ShutdownRoleOperation{
> +		OperationType: "ShutdownRoleOperation",
> +		Xmlns:         azureXmlns,

I don't think so. I'll remove all of them...


—
Reply to this email directly or view it on GitHub.

paulmey commented:
okay, wait with the merge, I found some bugs...

paulmey commented:
Fixed some namespace errors. Added the namespace to the XMLName property, which works as documented.
Also added another commit to pull apart the Deployment type for creating and looking at a deployment as they contain overlapping but distinct sets of properties.

paulmey commented:
Should be good to merge now. #65 builds on this and adds some integration test coverage.

ahmetalpbalkan commented:
@paulmey merged this, please rebase #65 so that it has newer commits on top of the merge commit.
